### ISSUE TITLE
HystrixRollingPercentile is producing wrong values

### ISSUE THREAD

**Issue by danielBreitlauch** [Netflix/Hystrix#1392]
The test below regularly adds values to the HystrixRollingPercentile, asks for the mean and sleeps afterwards. The value is increased until we sleep longer than the HystrixRollingPercentile moving window. At this point there wont be any values in the data structure because they are rotated out while we sleep.

If you run the code below you would expect to get increasing mean values until sleep is bigger than the rolling buckets window. Then the mean should be 0. After the sleep gets lower than the rolling window there should be a change in the mean again.

What happens is that it gets stuck at an arbitrary point and only returns this value. (In the example output it is 140)

Please increase cpuSpeedDependend if the test is running to fast on your cpu.

```
    public void testMeasuring() throws InterruptedException {
        int cpuSpeedDependend = 300;

        HystrixRollingPercentile executionDuration = new HystrixRollingPercentile(cpuSpeedDependend / 3, 10, 100, HystrixProperty.Factory.asProperty(true));

        for (int x=1; x < cpuSpeedDependend; x++) {
            executionDuration.addValue(x);
            int mean = executionDuration.getMean();
            System.out.println(x + ":" + mean);
            Thread.sleep(mean);
        }
        for (int x=cpuSpeedDependend; x > 1; x--) {
            executionDuration.addValue(x);
            int mean = executionDuration.getMean();
            System.out.println(x + ":" + mean);
            Thread.sleep(mean);
        }
    }
```

Output of the run:
...
16:8
17:8
18:9
19:9
20:10
...
98:97
99:98
100:99
101:101
102:0
103:0
104:0
105:0
...
175:0
176:0
177:0
178:0
179:140
180:140
181:140
182:140
...
300:140
299:140
...
5:140
4:140
3:140
2:140

**Comment by mattrjacobs** [comment]
Thanks for the report, @danielBreitlauch . Are you encountering this problem because you're using `HystrixRollingPercentile` outside of Hystrix, or via Hystrix?  I ask because Hystrix internally switched away from using this class in Version 1.5.0.

**Comment by danielBreitlauch** [comment]
I used it outside of Hystrix because it fits the use case perfectly. What is Hystrix using in 1.5.0?

**Comment by mattrjacobs** [comment]
There's an overview here: https://github.com/Netflix/Hystrix/wiki/Metrics-and-Monitoring.

Basically, each execution event gets put into an infinite stream of events, and then a number of stream consumers are set up to process these infinite stream to provide the same view of metrics as `HystrixRollingPercentile`.

A concrete example of one such consumer is: [RollingCommandEventCounterStream](https://github.com/Netflix/Hystrix/blob/master/hystrix-core/src/main/java/com/netflix/hystrix/metric/consumer/RollingCommandEventCounterStream.java)

**Comment by mattrjacobs** [comment]
I think that the fix for this is in #1407.  The problem was that advancing all buckets was not resetting the snapshot, causing it to appear to be stuck.  In practice, we've never observed this as having our total time tracked by the bucket array is much larger than 100ms.  

If you could apply that PR and try it out, I'd appreciate any feedback.

**Comment by danielBreitlauch** [comment]
I tested the fix and it looks good.
Thank you very much!

Btw. the 100ms window was just for testing. I encounter the bug with a 10s window too. Although it only happens when there is nothing happening for the whole timeframe.
